### PR TITLE
Tamanna/reduction keyframe extraction

### PR DESCRIPTION
**Pull request by tamannak8**
Adding keyframe extraction PR

### CODE REVIEW

File: reduction/keyframe_extraction/keyframe_extraction.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+def extract_frame_features(frame):
+    """
+        Extracts deep features from an input image frame using a pre-trained CNN model.
+
+        This function takes an image frame as input, preprocesses it by applying transformations
+        including resizing and normalization, and then extracts deep features from the image using
+        a pre-trained Convolutional Neural Network (CNN) model. The extracted features are flattened
+        and returned as a NumPy array.
+
+        Parameters:
```
Comment (vinayakSharm) [inline reduction/keyframe_extraction/keyframe_extraction.py:39]: ca you add the "------" underneath

File: reduction/keyframe_extraction/keyframe_extraction.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+def extract_key_frames(video_list, path="temp", num_key_frames=30):
+    """
+    Extracts and generates key frames from a list of input videos using K-Means clustering.
+
+    This function takes a list of video filenames, processes each video to extract frame features,
+    performs K-Means clustering to select representative key frames, and then generates a new video
+    containing the selected key frames. The generated video is further compressed using FFmpeg.
+
+    Args:
```
Comment (vinayakSharm) [inline reduction/keyframe_extraction/keyframe_extraction.py:72]: Parameters

**Review by vinayakSharm — COMMENTED** [review]
small docstring change, but looks good! good job

File: reduction/keyframe_extraction/keyframe_extraction.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
+import os
+import cv2
+from aEye import Aux
+import torch
+import torchvision.transforms as transforms
+from torchvision.models import resnet50
+import numpy as np
+from sklearn.cluster import KMeans
+import time
+import boto3
```
Comment (ankitadpatil) [inline reduction/keyframe_extraction/keyframe_extraction.py:10]: Unused import boto3

File: reduction/keyframe_extraction/keyframe_extraction.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    Returns:
+    ----------
+        None
+    """
+
+    for video in video_list:
+        video_path = os.path.join(video.get_file().strip("'"))
+        cap = cv2.VideoCapture(video_path)
+        video_name = Path(video_path).stem
+        frame_count = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
```
Comment (ankitadpatil) [inline reduction/keyframe_extraction/keyframe_extraction.py:91]: Unused variable frame_count

File: reduction/keyframe_extraction/keyframe_extraction.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        frame_width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
+        frame_height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
+        frame_rate = cap.get(cv2.CAP_PROP_FPS)
```
Comment (ankitadpatil) [inline reduction/keyframe_extraction/keyframe_extraction.py:94]: Unused variable frame_rate

File: reduction/keyframe_extraction/keyframe_extraction.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        # duration_seconds = frame_count / frame_rate
```
Comment (ankitadpatil) [inline reduction/keyframe_extraction/keyframe_extraction.py:95]: Please delete this line

**Review by ankitadpatil — CHANGES_REQUESTED** [review]

**Review by ankitadpatil — CHANGES_REQUESTED** [review]

File: reduction/keyframe_extraction/keyframe_extraction.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+from aEye import Aux
+import torch
+import torchvision.transforms as transforms
+from torchvision.models import resnet50
+import numpy as np
+from sklearn.cluster import KMeans
+import time
+from pathlib import Path
+import subprocess
+
```
[the remark this answers is not in the document]
Comment (ankitadpatil) [reply reduction/keyframe_extraction/keyframe_extraction.py:13]: The change has been added by @tamannak8

**Review by ankitadpatil — APPROVED** [review]
Everything looks and works well. Great job with the code for keyframe extraction technique!

**Pull request merged**